src/parse_execution.cpp

```diff
@@ -125,8 +125,8 @@ tnode_t<g::plain_statement> parse_execution_context_t::infinite_recursive_statem
     // Get the list of plain statements.
     // Ignore statements with decorations like 'builtin' or 'command', since those
     // are not infinite recursion. In particular that is what enables 'wrapper functions'.
-    tnode_t<g::statement> statement = first_job.child<1>();
-    tnode_t<g::job_continuation> continuation = first_job.child<2>();
+    tnode_t<g::statement> statement = first_job.child<2>();
+    tnode_t<g::job_continuation> continuation = first_job.child<3>();
```

Comment:
So these hardcoded child numbers seem quite suspect to me, but I never really got how our parser works.

This is because you have

```
[time] [command|builtin|...] therest
```

and so you're ignoring not just the `command|builtin` part, but also the `time` now? And time's separate because it can be stacked on top of the other decorators?

If so, the comment should be changed to indicate that.

Reply:
> So these hardcoded child numbers seem quite suspect to me, but I never really got how our parser works.

`tnode_t` provides functions that know about the grammar at compile time,
they can be used to catch errors at compile time, like when accessing a child that is out of bounds.

These here are fine because the values for `tnode_t::child` are given an
explicit type, so this will most likely fail to compile if a grammar
change invalidates the child number. (It is unlikely that two child
nodes have the same type.)

I had another place where I just used `tnode_t::child<1>().tag()`
(without specifying the nodes type, fixed now) - that is of dangerous
because the child type could change without the compiler noticing.

> And time's separate because it can be stacked on top of the other decorators?
> If so, the comment should be changed to indicate that.

Yes; the comment above those lines was misleading.
When accessing the `job_t`, we ignore prefixing variable assingments and `time`.
Non-plain statements, like `not` and other decorators are ignored later; I moved the comment down.